### ISSUE TITLE
plugin facility for sputnik collectors

### ISSUE THREAD

**Issue by sjamesr** [TouK/sputnik#113]
Hi there,

I'd like to annotate our Go reviews with the output of [Go Vet](http://golang.org/cmd/vet/). As I understand it, I'd have to provide a new collector implementation that invokes Go Vet on the before- and after-review state of the Git repo.

I'm wondering if there's any interest in implementing a plugin system for Sputnik. I imagine it would be useful if Sputnik could locate new collector implementations at runtime. I think this would make it a little easier to introduce new collector types without changing the Sputnik source code.

Thanks,
James

**Comment by zygm0nt** [comment]
This would be a very welcome change. So if you'd like to contribute, than please do so.

Thanks.

**Comment by mirkosertic** [comment]
I think this would be very cool, as i need this feature too. I will try to add this, shouln't be such a big deal...

**Comment by pjagielski** [comment]
I don't think that executing external binaries by Sputnik is a good idea. Is this what you mean by invoking Go Vet?

What we were thinking of instead could be a set of plugins that convert other tools output to Sputnik-compatible input. So the tool would be run outside Sputnik and then the output could be passed to Sputnik [ReviewProcessor](https://github.com/TouK/sputnik/blob/80bbfb4c396a08a04d3edb402597bee9ffdf240d/src/main/java/pl/touk/sputnik/review/ReviewProcessor.java) to construct [ReviewResult](https://github.com/TouK/sputnik/blob/80bbfb4c396a08a04d3edb402597bee9ffdf240d/src/main/java/pl/touk/sputnik/review/ReviewResult.java). Sputnik can then send the result to Gerrit/Stash.

Is this reasonable?
